OrbitService/CaptureServiceImpl.cpp

```diff
@@ -146,13 +150,37 @@ grpc::Status CaptureServiceImpl::Capture(
   }
   LOG("Client finished writing on Capture's gRPC stream: stopping capture");
 
-  tracing_handler.Stop();
-  LOG("LinuxTracingHandler stopped: perf_event_open tracing is done");
+  {
+    std::vector<std::thread> stop_threads;
+    stop_threads.emplace_back([&tracing_handler] {
+      tracing_handler.Stop();
+      LOG("LinuxTracingHandler stopped: perf_event_open tracing is done");
+    });
+    for (CaptureStartStopListener* listener : capture_start_stop_listeners_) {
+      stop_threads.emplace_back([&listener] {
+        listener->OnCaptureStopRequested();
+        LOG("CaptureStartStopListener stopped: one or more producers finished capturing");
+      });
+    }
+    for (std::thread& stop_thread : stop_threads) {
+      stop_thread.join();
+    }
+  }
 
   capture_event_buffer.StopAndWait();
   LOG("Finished handling gRPC call to Capture: all capture data has been sent");
   is_capturing = false;
   return grpc::Status::OK;
 }
 
+void CaptureServiceImpl::AddCaptureStartStopListener(CaptureStartStopListener* listener) {
```

Comment:
Would that be the right point in time to start writing tests for this class?

Reply:
Which class do you mean? `CaptureServiceImpl`? I'm really not sure how to do that, as it's a gRPC service. Unless we involve gRPC in the tests, but should we go there?

Reply:
Yes, I was referring to `CaptureServiceImpl`. Also not sure if we should go this route. Could mock away the gRPC stuff here?
Also feel free, to just ignore that comment.

Reply:
We should consider that but it's out of scope here. I guess `grpc::ServerReaderWriter<CaptureResponse, CaptureRequest>` could be mocked and then `Capture` can be called locally. `LinuxTracingHandler` would also need to be mocked. Maybe it can also inherit from `CaptureStartStopListener`.

Reply:
Agree. Should we create a bug for this?

Reply:
http://b/174145842

Reply:
Thanks!